### PR TITLE
expression,planner: support non-deterministic functions (e.g., now) in the prepared plan cache

### PR DESCRIPTION
**Pull request by dbjoa**
### What problem does this PR solve? 
Support non-deterministic functions (e.g., now) in the prepared plan cache. 

### What is changed and how it works?
Make a non-deterministic function expression to be a constant expression with the deferred expression (i.e., the non-deterministic function expression).

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change

Side effects

 - No

Related changes

 - Need to be included in the release note

### CODE REVIEW

**Comment by shenli** [comment]
@dbjoa Thanks!

File: expression/scalar_function.go [hunk 1]
```diff
@@ -70,8 +70,8 @@ func (sf *ScalarFunction) MarshalJSON() ([]byte, error) {
 	return []byte(fmt.Sprintf("\"%s\"", sf)), nil
 }
 
-// NewFunction creates a new scalar function or constant.
-func NewFunction(ctx sessionctx.Context, funcName string, retType *types.FieldType, args ...Expression) (Expression, error) {
+// newFunction creates a new scalar function or constant.
```
[the remark this answers is not in the document]
Comment (dbjoa) [reply expression/scalar_function.go]: Thank you for the comment. I've updated the PR.

File: planner/core/prepare_test.go [hunk 2]
```diff
@@ -15,9 +15,13 @@ package core_test
 
 import (
 	. "github.com/pingcap/check"
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/metrics"
 	"github.com/pingcap/tidb/planner/core"
 	"github.com/pingcap/tidb/util/testkit"
 	"github.com/pingcap/tidb/util/testleak"
+	dto "github.com/prometheus/client_model/go"
+	"time"
```
Comment (XuHuaiyu) [inline planner/core/prepare_test.go]: put this line to line 17 and leave a blank line behind it.

File: expression/scalar_function.go [hunk 3]
```diff
@@ -70,8 +70,8 @@ func (sf *ScalarFunction) MarshalJSON() ([]byte, error) {
 	return []byte(fmt.Sprintf("\"%s\"", sf)), nil
 }
 
-// NewFunction creates a new scalar function or constant.
-func NewFunction(ctx sessionctx.Context, funcName string, retType *types.FieldType, args ...Expression) (Expression, error) {
+// newFunctionImpl creates a new scalar function or constant.
+func newFunctionImpl(fold bool, ctx sessionctx.Context, funcName string, retType *types.FieldType, args ...Expression) (Expression, error) {
```
Comment (XuHuaiyu) [inline expression/scalar_function.go]: `ctx` is always the first argument of a function,
we'd better keep this style.

**Comment by XuHuaiyu** [comment]
MySQL:
``` sql
mysql> prepare stmt from "select now(), sleep(1), now(), sleep(1), now()";
Query OK, 0 rows affected (0.00 sec)
Statement prepared

mysql> execute stmt;
+---------------------+----------+---------------------+----------+----------------$----+
| now()               | sleep(1) | now()               | sleep(1) | now()
    |
+---------------------+----------+---------------------+----------+----------------$----+
| 2018-10-31 16:00:14 |        0 | 2018-10-31 16:00:14 |        0 | 2018-10-31 16:0$:14 |
+---------------------+----------+---------------------+----------+----------------$----+
1 row in set (2.01 sec)

mysql> execute stmt;
+---------------------+----------+---------------------+----------+----------------$----+
| now()               | sleep(1) | now()               | sleep(1) | now()
    |
+---------------------+----------+---------------------+----------+----------------$----+
| 2018-10-31 16:00:19 |        0 | 2018-10-31 16:00:19 |        0 | 2018-10-31 16:0$:19 |
+---------------------+----------+---------------------+----------+----------------$----+
1 row in set (2.01 sec)
```

In MySQL, `now` is a foldable function,
but It seems we will get different results for different `now()` with this commit?

**Comment by dbjoa** [comment]
@XuHuaiyu, Thank you for the comment. The updated PR should address the issue.

File: planner/core/prepare_test.go (hunk 2 above)
Comment (dbjoa) [reply planner/core/prepare_test.go]: Thank you for the detailed comment. I've updated the PR.

File: expression/scalar_function.go (hunk 3 above)
Comment (dbjoa) [reply expression/scalar_function.go]: Thank you for the detailed comment. I've updated the PR.

**Comment by dbjoa** [comment]
I should also handle the other functions like ``NOW()``.  (i.e., ``CURDATE()``, ``CURTIME()``, ``UTC_DATE()``, ``UTC_TIME()``, ``UTC_TIMESTAMP()``, ``UNIX_TIMESTAMP()``)

File: planner/core/prepare_test.go [hunk 4]
```diff
@@ -14,7 +14,13 @@
 package core_test
 
 import (
+	dto "github.com/prometheus/client_model/go"
```
Comment (zz-jason) [inline planner/core/prepare_test.go]: please move this package to the lower block. separate it from the golang standard packages.

File: expression/function_traits.go [hunk 5]
```diff
@@ -52,6 +41,23 @@ var unFoldableFunctions = map[string]struct{}{
 	ast.GetParam:  {},
 }
 
+// DeferredFunctions stores non-deterministic functions which can be deferred.
```
Comment (zz-jason) [inline expression/function_traits.go]: How adding another comment to specify that the deferred functions take effects only when plan cache is enabled?

File: planner/core/prepare_test.go (hunk 4 above)
Comment (dbjoa) [reply planner/core/prepare_test.go]: Thank you for the comment. I've updated the PR.

File: expression/function_traits.go (hunk 5 above)
Comment (dbjoa) [reply expression/function_traits.go]: Thank you for the comment. I've added the comment to the updated the PR.

**Pull request merged**